### PR TITLE
Adding scatter_to/scatter_from collective operations

### PR DESCRIPTION
**Pull request by hkaiser**
- ~~adding hpx::gather API combining gather_here/gather_there~~ (adding this was a mistake as it did turn gather into an equivalent for all_gather)
- adding missing gather test
- fixing various problems and inconsistencies with broadcast

### CODE REVIEW

**Comment by hkaiser** [comment]
This is ready to go now (once the test have cycled), please review.

File: libs/collectives/include/hpx/collectives/scatter.hpp [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+    {
+        if (num_sites == std::size_t(-1))
+        {
+            num_sites = static_cast<std::size_t>(
+                hpx::get_num_localities(hpx::launch::sync));
+        }
+        if (this_site == std::size_t(-1))
+        {
+            this_site = static_cast<std::size_t>(hpx::get_locality_id());
+        }
```
Comment (msimberg) [inline libs/collectives/include/hpx/collectives/scatter.hpp:417]: Is there a reason these can't just directly be the default parameters? I suppose to be able to specify `generation` but leave `num_sites` default? Related question, but definitely not for this PR: should `get_num_localities` and the locality id just be `std::size_t`?

Comment (hkaiser) [reply libs/collectives/include/hpx/collectives/scatter.hpp:417]: > Is there a reason these can't just directly be the default parameters? I suppose to be able to specify generation but leave num_sites default?

Yes, this is fixed by #4687 (which relies on this PR).

> Related question, but definitely not for this PR: should get_num_localities and the locality id just be std::size_t?

Ok, not objections. See #4697.

**Comment by khuck** [comment]
@hkaiser I think this merge is the one that broke APEX.  What is the migration path for replacing `hpx::lcos::gather_here()` and friends?  Is there any way to add an APEX build to the tests so that this would get caught sooner?

I tried replacing `hpx::lcos::gather_here` with `hpx::collectives::gather_here()` but that didn't work... same with `gather_there`, `broadcast_to` and `broadcast_from`.

**Comment by hkaiser** [comment]
> @hkaiser I think this merge is the one that broke APEX. What is the migration path for replacing `hpx::lcos::gather_here()` and friends? Is there any way to add an APEX build to the tests so that this would get caught sooner?

Uhh, sorry - I was not aware that you were relying on those functions :/ 

> I tried replacing `hpx::lcos::gather_here` with `hpx::collectives::gather_here()` but that didn't work... same with `gather_there`, `broadcast_to` and `broadcast_from`.

The functions have been moved to the namespace `hpx::collectives` (as you suspected). We have however introduced special types for the various arguments to prevent mixing the various `size_t` values. Please see https://github.com/STEllAR-GROUP/hpx/blob/master/libs/full/collectives/tests/unit/broadcast.cpp and https://github.com/STEllAR-GROUP/hpx/blob/master/libs/full/collectives/tests/unit/gather.cpp for examples.

**Comment by khuck** [comment]
@hkaiser When I change the code to match the example, I get this error message now (I am reducing std::string objects, if that matters):

```
Scanning dependencies of target apex
[ 72%] Building CXX object apex/src/apex/CMakeFiles/apex.dir/otf2_listener.cpp.o
In file included from /scratch/khuck/hpx/build/libs/core/threading_base/include/hpx/modules/threading_base.hpp:14:0,
                 from /scratch/khuck/hpx/src/libs/core/synchronization/include/hpx/synchronization/mutex.hpp:14,
                 from /scratch/khuck/hpx/src/libs/core/synchronization/include/hpx/synchronization/condition_variable.hpp:16,
                 from /scratch/khuck/hpx/src/libs/parallelism/futures/include/hpx/futures/detail/future_data.hpp:20,
                 from /scratch/khuck/hpx/src/libs/parallelism/async_combinators/include/hpx/async_combinators/wait_all.hpp:108,
                 from /scratch/khuck/hpx/src/libs/parallelism/execution/include/hpx/execution/executors/execution.hpp:17,
                 from /scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/execution_policy.hpp:14,
                 from /scratch/khuck/hpx/src/libs/parallelism/algorithms/include/hpx/parallel/algorithms/adjacent_find.hpp:124,
                 from /scratch/khuck/hpx/src/libs/parallelism/algorithms/include/hpx/parallel/algorithm.hpp:17,
                 from /scratch/khuck/hpx/src/libs/full/include_local/include/hpx/local/algorithm.hpp:10,
                 from /scratch/khuck/hpx/src/libs/full/include/include/hpx/algorithm.hpp:10,
                 from /scratch/khuck/hpx/src/libs/full/include/include/hpx/hpx.hpp:10,
                 from /scratch/khuck/hpx/build/_deps/apex-src/src/apex/otf2_listener.cpp:11:
/scratch/khuck/hpx/src/libs/core/threading_base/include/hpx/threading_base/annotated_function.hpp:72:26 warning: ‘warn_unused_result’ attribute only applies to function types [-Wattributes]
     struct HPX_NODISCARD annotate_function
                          ^~~~~~~~~~~~~~~~~
In file included from /scratch/khuck/hpx/src/libs/parallelism/async_local/include/hpx/async_local/dataflow.hpp:11:0,
                 from /scratch/khuck/hpx/src/libs/parallelism/algorithms/include/hpx/parallel/util/partitioner.hpp:12,
                 from /scratch/khuck/hpx/src/libs/parallelism/algorithms/include/hpx/parallel/algorithms/adjacent_find.hpp:132,
                 from /scratch/khuck/hpx/src/libs/parallelism/algorithms/include/hpx/parallel/algorithm.hpp:17,
                 from /scratch/khuck/hpx/src/libs/full/include_local/include/hpx/local/algorithm.hpp:10,
                 from /scratch/khuck/hpx/src/libs/full/include/include/hpx/algorithm.hpp:10,
                 from /scratch/khuck/hpx/src/libs/full/include/include/hpx/hpx.hpp:10,
                 from /scratch/khuck/hpx/build/_deps/apex-src/src/apex/otf2_listener.cpp:11:
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp: In instantiation of ‘struct hpx::lcos::detail::dataflow_return_impl<false, hpx::detail::sync_policy, hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>, hpx::tuple<hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, void>’:
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:140:12:   required from ‘struct hpx::lcos::detail::dataflow_return<hpx::detail::sync_policy, hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>, hpx::tuple<hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >’
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:164:12:   required from ‘struct hpx::lcos::detail::dataflow_frame<hpx::detail::sync_policy, hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>, hpx::tuple<hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >’
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:414:26:   required by substitution of ‘template<class Allocator, class Policy, class Func, class ... Ts, class Frame> typename Frame::type hpx::lcos::detail::create_dataflow_alloc(const Allocator&, Policy&&, Func&&, Ts&& ...) [with Allocator = std::allocator<int>; Policy = const hpx::detail::sync_policy&; Func = hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>; Ts = {hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; Frame = hpx::lcos::detail::dataflow_frame<hpx::detail::sync_policy, hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>, hpx::tuple<hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >]’
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:445:49:   required from ‘static decltype(auto) hpx::lcos::detail::dataflow_dispatch_impl<false, Policy, typename std::enable_if<hpx::traits::is_launch_policy<Policy>::value>::type>::call(const Allocator&, Policy_&&, F&&, Ts&& ...) [with Allocator = std::allocator<int>; Policy_ = const hpx::detail::sync_policy&; F = hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>; Ts = {hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; Policy = hpx::detail::sync_policy]’
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:458:68:   required by substitution of ‘template<class Allocator, class F, class ... Ts> static decltype (hpx::lcos::detail::dataflow_dispatch_impl<false, hpx::detail::sync_policy, void>::call<Allocator, Policy_, F, Ts ...>(alloc, forward<F>(f), (forward<Ts>)(hpx::lcos::detail::dataflow_dispatch<Policy, typename std::enable_if<hpx::traits::is_launch_policy::value, void>::type>::call::ts)...)) hpx::lcos::detail::dataflow_dispatch<hpx::detail::sync_policy, void>::call<Allocator, F, Ts ...>(const Allocator&, F&&, Ts&& ...) [with Allocator = std::allocator<int>; F = const hpx::detail::sync_policy&; Ts = {hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)>, hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}]’
/scratch/khuck/hpx/src/libs/parallelism/async_base/include/hpx/async_base/dataflow.hpp:25:76:   required by substitution of ‘template<class F, class ... Ts> decltype (hpx::lcos::detail::dataflow_dispatch<typename std::decay<_Tp>::type>::call(hpx::util::internal_allocator<>{}, forward<F>(f), (forward<Ts>)(hpx::dataflow::ts)...)) hpx::dataflow(F&&, Ts&& ...) [with F = const hpx::detail::sync_policy&; Ts = {hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)>, hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}]’
/scratch/khuck/hpx/src/libs/full/collectives/include/hpx/collectives/gather.hpp:272:24:   required from ‘hpx::lcos::future<std::vector<typename std::decay<_Tp>::type> > hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]’
/scratch/khuck/hpx/src/libs/full/collectives/include/hpx/collectives/gather.hpp:282:27:   required from ‘hpx::lcos::future<std::vector<typename std::decay<_Tp>::type> > hpx::collectives::gather_here(const char*, T&&, hpx::collectives::num_sites_arg, hpx::collectives::this_site_arg, hpx::collectives::generation_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]’
/scratch/khuck/hpx/build/_deps/apex-src/src/apex/otf2_listener.cpp:1422:45:   required from here
/scratch/khuck/hpx/src/libs/parallelism/executors/include/hpx/executors/dataflow.hpp:116:71: error: no type named ‘type’ in ‘struct hpx::util::detail::invoke_fused_result<hpx::collectives::gather_here(hpx::collectives::communicator, T&&, hpx::collectives::this_site_arg) [with T = std::__cxx11::basic_string<char>&; typename std::decay<_Tp>::type = std::__cxx11::basic_string<char>]::<lambda(hpx::collectives::communicator&&, std::__cxx11::basic_string<char>&)>, hpx::tuple<hpx::collectives::communicator, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >’
             typename util::detail::invoke_fused_result<F, Args>::type>;
                                                                       ^
```

The code there looks like this now:

```C++
1394     std::unique_ptr<std::tuple<std::map<int,int>,
1395                     std::map<int,std::string> > >
1396          otf2_listener::reduce_node_properties(std::string&& str) {
1397         using namespace hpx::collectives;
1398         static std::uint32_t generation{0};
1399         std::uint32_t num_localities = hpx::get_num_localities(hpx::launch::sync);
1400         std::uint32_t this_locality = hpx::get_locality_id();
1401         // get all hostnames
1402         constexpr char const* gather_basename = "/otf2/properties/gather_direct/";
1403 
1404         // if not root, we have a simple job...
1405         if (my_saved_node_id > 0) {
1406             //printf("%d: calling gather_there from %s\n", my_saved_node_id, __func__);
1407             hpx::future<void> overall_result =
1408                 gather_there(gather_basename, str,
1409                 this_site_arg(this_locality),
1410                 generation_arg(generation++));
1411             overall_result.get();
1412             return nullptr;
1413         }
1414         std::vector<std::string> allhostnames;
1415         if (my_saved_node_count > 1) {
1416             //printf("%d: calling gather_here from %s\n", my_saved_node_id, __func__);
1417             // if root, gather the names...
1418             hpx::future<std::vector<std::string>> overall_result =
1419                 gather_here(gather_basename, str,
1420                 num_sites_arg(num_localities),
1421                 this_site_arg(this_locality),
1422                 generation_arg(generation++));
1423             allhostnames = overall_result.get();
1424         } else {
1425             allhostnames.push_back(str);
1426         }
...
```

**Comment by hkaiser** [comment]
@khuck: I can confirm this problem. Sometimes it makes sense to write exhaustive tests after all... Please see #5388 for a possible fix.

**Comment by msimberg** [comment]
> Is there any way to add an APEX build to the tests so that this would get caught sooner?

Just so I understand what's going on here, did this break with APEX master/develop or with the latest release? Because we do have a CI configuration with APEX, but it only uses the tagged version.

**Comment by hkaiser** [comment]
> > Is there any way to add an APEX build to the tests so that this would get caught sooner?
> 
> Just so I understand what's going on here, did this break with APEX master/develop or with the latest release? Because we do have a CI configuration with APEX, but it only uses the tagged version.

I introduced a breaking change to the collective APIs assuming nobody was using them anyways. APEX was using them, apparently, so the change broke it (no idea why it didn't show up on the APEX builder). Thanks @khuck for your patience with this - I should have been more careful.

**Comment by msimberg** [comment]
> (no idea why it didn't show up on the APEX builder)

My question was exactly about this. @khuck in what configuration and with what versions did your build fail? This PR was merged over a year ago which is why I'm surprised that this came up only now.

**Comment by khuck** [comment]
This breaks latest tag 2.4.0 and current develop branch.  If you don't build APEX with OTF2 support, this code isn't compiled.  The event unification happens at the end of execution - the OTF2 library relies on the application's communication infrastructure to make a global list of timer names, counter names, and other strings.  I'll merge my changes and make a 2.4.1 tag for this fix.  We'll want to update the tag that HPX uses as well.

**Comment by hkaiser** [comment]
> > (no idea why it didn't show up on the APEX builder)
> 
> My question was exactly about this. @khuck in what configuration and with what versions did your build fail? This PR was merged over a year ago which is why I'm surprised that this came up only now.

FWIW, the breaking changes were introduced recently only (#5357).

**Comment by msimberg** [comment]
Thanks @khuck and @hkaiser for the explanations. It's the OTF2 support that is missing from the CI configuration. I'll add that.

**Pull request merged**